### PR TITLE
DOCSUP-18731: Add CustomSeparatedWithNames and CustomSeparatedWithNamesAndTypes formats

### PR DESCRIPTION
**Pull request by sevirov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:

- Documentation for #31434

### CODE REVIEW

File: docs/en/operations/settings/settings.md [hunk 1]
```diff
@@ -4071,3 +4071,55 @@ Possible values:
...
+-   `'CSV'` — Similarly to [CSV](../../interfaces/formats.md#csv).
+-   `'JSON'` — Similarly to [JSONEachRow](../../interfaces/formats.md#jsoneachrow).
+-   `'XML'` — Similarly to [XML](../../interfaces/formats.md#xml).
+-   `'Raw'` — Extracts subpatterns as a whole, no escaping rules.
+
+Default value: `'Escaped'`.
+
+## format_custom_field_delimiter {#format-custom-field-delimiter}
+
+Sets the character is interpreted as a delimiter between fields for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```
Comment (gyuton) [inline docs/en/operations/settings/settings.md:4093]:
```suggestion
Sets the character that is interpreted as a delimiter between fields for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```

File: docs/en/operations/settings/settings.md [hunk 2]
```diff
@@ -4071,3 +4071,55 @@ Possible values:
...
+
+Default value: `'\t'`.
+
+## format_custom_row_before_delimiter {#format-custom-row-before-delimiter}
+
+Sets the character is interpreted as a delimiter before field of the first column for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```
Comment (gyuton) [inline docs/en/operations/settings/settings.md:4099]:
```suggestion
Sets the character that is interpreted as a delimiter before field of the first column for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```

File: docs/en/operations/settings/settings.md [hunk 3]
```diff
@@ -4071,3 +4071,55 @@ Possible values:
...
+
+Default value: `''`.
+
+## format_custom_row_after_delimiter {#format-custom-row-after-delimiter}
+
+Sets the character is interpreted as a delimiter after field of the last column for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```
Comment (gyuton) [inline docs/en/operations/settings/settings.md:4105]:
```suggestion
Sets the character that is interpreted as a delimiter after the field of the last column for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```

File: docs/en/operations/settings/settings.md [hunk 4]
```diff
@@ -4071,3 +4071,55 @@ Possible values:
...
+
+Default value: `'\n'`.
+
+## format_custom_row_between_delimiter {#format-custom-row-between-delimiter}
+
+Sets the character is interpreted as a delimiter between rows for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```
Comment (gyuton) [inline docs/en/operations/settings/settings.md:4111]:
```suggestion
Sets the character that is interpreted as a delimiter between rows for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```

File: docs/en/operations/settings/settings.md [hunk 5]
```diff
@@ -4071,3 +4071,55 @@ Possible values:
...
+
+Default value: `''`.
+
+## format_custom_result_before_delimiter {#format-custom-result-before-delimiter}
+
+Sets the character is interpreted as a prefix before result set for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```
Comment (gyuton) [inline docs/en/operations/settings/settings.md:4117]:
```suggestion
Sets the character that is interpreted as a prefix before the result set for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```

File: docs/en/operations/settings/settings.md [hunk 6]
```diff
@@ -4071,3 +4071,55 @@ Possible values:
...
+
+Default value: `''`.
+
+## format_custom_result_after_delimiter {#format-custom-result-after-delimiter}
+
+Sets the character is interpreted as a suffix after result set for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```
Comment (gyuton) [inline docs/en/operations/settings/settings.md:4123]:
```suggestion
Sets the character that is interpreted as a suffix after the result set for [CustomSeparated](../../interfaces/formats.md#format-customseparated) data format.
```

**Comment by sevirov** [comment]
@Avogar Please, review this PR.

File: docs/en/interfaces/formats.md [hunk 7]
```diff
@@ -23,6 +23,8 @@ The supported formats are:
 | [CSV](#csv)                                                                             | ✔     | ✔      |
 | [CSVWithNames](#csvwithnames)                                                           | ✔     | ✔      |
 | [CSVWithNamesAndTypes](#csvwithnamesandtypes)                                                           | ✔     | ✔      |
+| [CustomSeparatedWithNames](#customseparatedwithnames)                                   | ✔     | ✔      |
+| [CustomSeparatedWithNamesAndTypes](#customseparatedwithnamesandtypes)                   | ✔     | ✔      |
 | [CustomSeparated](#format-customseparated)                                              | ✔     | ✔      |
```
Comment (Avogar) [inline docs/en/interfaces/formats.md:28]: Let's add them after `CustomSeparated`

File: docs/en/interfaces/formats.md [hunk 8]
```diff
@@ -1536,14 +1547,19 @@ Each line of imported data is parsed according to the regular expression.
...
+
+-   `format_regexp_escaping_rule` — [String](../sql-reference/data-types/string.md). The following escaping rules are supported:
+
+    -   CSV (similarly to [CSV](#csv))
+    -   JSON (similarly to [JSONEachRow](#jsoneachrow))
+    -   Escaped (similarly to [TSV](#tabseparated))
+    -   Quoted (similarly to [Values](#data-format-values))
+    -   Raw (extracts subpatterns as a whole, no escaping rules)
+    -   XML (similarly to [XML](#xml))
+    -   None (no escaping rules)
```
Comment (Avogar) [inline docs/en/interfaces/formats.md:1560]: XML and None are not supported in Regexp

File: docs/en/operations/settings/settings.md [hunk 9]
```diff
@@ -4071,3 +4071,55 @@ Possible values:
...
+
+Possible values:
+
+-   `'None'` — No escaping rules.
+-   `'Escaped'` — Similarly to [TSV](../../interfaces/formats.md#tabseparated).
+-   `'Quoted'` — Similarly to [Values](../../interfaces/formats.md#data-format-values).
+-   `'CSV'` — Similarly to [CSV](../../interfaces/formats.md#csv).
+-   `'JSON'` — Similarly to [JSONEachRow](../../interfaces/formats.md#jsoneachrow).
+-   `'XML'` — Similarly to [XML](../../interfaces/formats.md#xml).
+-   `'Raw'` — Extracts subpatterns as a whole, no escaping rules.
```
Comment (Avogar) [inline docs/en/operations/settings/settings.md:4087]: For `Raw` we can add that it similar to `TSVRaw`

Comment (Avogar) [reply docs/en/operations/settings/settings.md:4087]: And let's remove `None` since it make sense only for Template

**Review by Avogar — APPROVED** [review]

**Pull request merged**